### ISSUE TITLE
Temporal Filtering of Features is wrong

### ISSUE THREAD

**Issue by rduivenvoorde** [qgis/QGIS#41970]
@nyalldawson @timlinux @anitagraser @Samweli 

My really last try to make my point about why temporal filtering of features is wrong when the data consists of time-intervals.
Using terms from https://grass.osgeo.org/grass79/manuals/t.vect.algebra.html here.

This ZIP: 
[tempdemo.zip](https://github.com/qgis/QGIS/files/6068525/tempdemo.zip)
contains a shp file and a csv with time fields and 3 project files to play with.
Every project file defines the temporal properties in the 4 different ways you are able to define a time-intervals and 1 way to define a time-instance

Note 1: in case the temporal property is a time-instance (NO duration) all is fine, QGIS uses the "left closed right open interval" rule

Note 2: I partialy fixed the handling of time-intervals in:
https://github.com/qgis/QGIS/pull/39568
(BUT ONLY in the case you define your temporal time-interval using a single field and a fixed duration)

Note 3: I think/hope all is fixed with https://github.com/qgis/QGIS/pull/40989

To show what goes wrong: below we show the selecting of data (time-intervals) with a time-interval (Frame) of one hour.
In my opinion you should ONLY see the 02:00:00-03:00:00 data (current time/view Frame):

![Screenshot-20210302125208-895x305](https://user-images.githubusercontent.com/731673/109644709-213d4700-7b56-11eb-8b06-f2718a29bd1d.png)

Only the case of the 'single field and fixed 1hr duration' it is behaving right (as fixed in above mentioned PR)

This is another example: we select the data (with 1 hour time-intervals) with a time-frame of one hour.
Both data time-intervals and filter-time-interval (Frame)'s are 'equal'.
In my opinion the preceding and following interval should not be taken into account.
So looking at data of 8:00-9:00 you should NOT see 7:00-8:00 (which you do) in my opinion.
Looking at the t.vect.algebra.html rules ("left closed right open interval") you could argument that we ONLY want the 'equals' relation and not preceding or finishing though you could argue about this because of the "left closed right open interval" rule:

![Screenshot-20210302125534-1702x602](https://user-images.githubusercontent.com/731673/109645056-9a3c9e80-7b56-11eb-8b63-fa23ba0c7666.png)

Please test/try/comment

**Comment by jakimowb** [comment]
I was able to visualize it as expected by ensuring none-overlapping intervalls (t_start2 = t_start + 1 sec, t_end2 = t_end - 1 sec)
It would be great to have a user's choice on how data intervalls are interpreted, e.g. left open/closed and right open/closed.

![grafik](https://user-images.githubusercontent.com/1404870/111596034-05c87200-87cd-11eb-8de1-4ca1c0992bc8.png)


I realized that I cannot simply use the expression builder to "add" or "remove" 1 second from a datetime value (without string manipulations). Otherwise, t_start and t_end could be simply adjusted in the layer properties:
![grafik](https://user-images.githubusercontent.com/1404870/111598161-33aeb600-87cf-11eb-8cfe-360dc2946e3e.png)
